### ISSUE TITLE
Should not_null have conversion operators?

### ISSUE THREAD

**Issue by petamas** [gsl-lite/gsl-lite#253]
There are a few problems with `not_null`'s current conversions:
- conversion operators are evil by default, having to be explicit by using a function helps to avoid bugs
- the current implementation can't convert `not_null<unique_ptr<T>>` to `unique_ptr<T>` if `FUNCTION_REF_QUALIFIER` feature is missing from the compiler (eg. GCC 4.7.3)
- if someone uses MSVC 2017 locally, but has GCC 4.7.3 in their CI system, then MSVC lets them build code which cannot be built with GCC, resulting in later detection of issues
- it's really, really hard to read the relevant part of gsl-lite.hpp: we have 5 different conversion operator variants in code, `#ifdef`ed to the moon and back (naturally, this mostly impacts the maintainers, not the users, but it's a problem still)

Using `as_nullable` (see PR#251) instead of conversion operators alleviates all of these problems.

I have two proposals:
1. Introduce a gsl_CONFIG variable to enable/disable the conversion operators - users who need to always use `as_nullable` because of GCC 4.7.3 compatibility can turn the operators off in all their builds, helping them catch problems during local development.
2. In a later version, switch the config to "off", or remove it entirely - all code will become better if they use the explicit `as_nullable` functions.

What do you think about these changes?

**Comment by mbeutel** [comment]
First and foremost, I think we can all agree that `not_null<>` is an abomination. Our implementation is very difficult to comprehend as it tries to satisfy as many conflicting demands as possible, but I believe there generally isn't much room for improvement unless we do away with a bunch of design goals (such as movability). C++ just isn't the right language for expressing this kind of guarantee, and no amount of overloading and specializing will make an implementation of `not_null<>` safe; and of course our support for legacy compilers makes it worse.


> 1. Introduce a gsl_CONFIG variable to enable/disable the conversion operators - users who need to always use `as_nullable` because of GCC 4.7.3 compatibility can turn the operators off in all their builds, helping them catch problems during local development.

I'd be fine with a configuration option for disabling `not_null<>` conversion operators.


> 2. In a later version, switch the config to "off", or remove it entirely - all code will become better if they use the explicit `as_nullable` functions.

I share your general sentiments about conversion operators. Also, I sympathize with your legacy MSVC/legacy GCC scenario, and I'd be happy to add a config switch if that makes it easier to bear. But I can't follow here: why does code become better when not relying on this implicit conversion? You argue convincingly that conversion operators are a burden to maintainers; but how do these particular conversion operators make user code worse?

In my opinion, choosing a different version-1 default for this configuration option would necessitate an inquiry with the Guidelines maintainers. The implicit conversions to the underlying type are sound from a type system perspective (unlike the implicit constructors *from* the underlying type which M-GSL omits in its `strict_not_null<>` and which will be disabled by default in *gsl-lite* version 1), and Microsoft's *de-facto* standard GSL implementation also has conversion operators in both `not_null<>` and `strict_not_null<>`.

I'm not saying we cannot deviate from the Guidelines specification here if there are compelling reasons to do so, but we should at least have upstream discussions with the Guidelines and M-GSL folks.

(Perhaps a more general upstream discussion of `strict_not_null<>` + `not_null<>` vs. `not_null<>` + `not_null_ic<>` would also be desirable.)

**Comment by petamas** [comment]
Having conversion operators and conversion constructors can cause subtle issues, which in my opinion should be avoided as much as possible. In #257, I describe such a scenario; it is not 100% relevant here, as it is about the other direction (converting nullable to `not_null`), but I'm pretty sure a similar (but a bit more contrived) example could be constructed for this direction too.

Being able to simply pass a `not_null<T>` as a `T` to some function seems useful at first, but after introducing `not_null` in our codebase, I found that most things wrapped once in `not_null` basically never get unwrapped. Almost every time we had to use `as_nullable` (because of GCC 4.7.3) to pass a `not_null<T>` as a `T`, we discovered an interface that could not handle NULLs, so it should have consumed a `not_null<T>` in teh first place, eliminating the need for the conversion. This way, `as_nullable` became a marker for the borders between the "GSL-ed" and "non-GSL-ed" parts of the codebase; the remaining few cases where the consumer _could_ handle NULLs, but we wanted to pass a `not_null` were simply commented.

TLDR: other than my general disdain for silent conversions, I think there are only a few places where conversion is needed, and it's better to mark these places with an easily-searched, "ugly" function instead of letting it silently do its magic. (Kind of similarly to how static_cast etc. are preferred to C-style casts, or how manually constructing a unique_ptr from a raw pointer signals that someone passed you a raw pointer instead of a smart one.)

> legacy MSVC/legacy GCC scenario

The problem is "modern MSVC / legacy GCC" - with a legacy MSVC, neither would compile the "advanced" code, so there would be no problem. :)

Option 1 would solve my problems, the 2nd proposal's more like a "what-if" scenario. I do not want to make everyone in the world conform to my style/standards, but I'm always wondering if others would benefit from them. If you prefer to have the conversion by default in the long term, it's fine by me. (But then I have to modify the test suite to include tests with both values of the config. :D)

**Comment by petamas** [comment]
> (Perhaps a more general upstream discussion of strict_not_null<> + not_null<> vs. not_null<> + not_null_ic<> would also be desirable.)

Not surprisingly, I prefer GSL-Lite's version, having an implicit constructor defeats the whole purpose of `not_null` in my eyes. :)

By "upstream discussion", do you mean discussion with the M-GSL folks, discussion with the Core Guidelines people, or discussion among ourselves about perhaps renaming our classes to match M-GSL?

**Comment by mbeutel** [comment]
> Having conversion operators and conversion constructors can cause subtle issues, which in my opinion should be avoided as much as possible. In #257, I describe such a scenario; it is not 100% relevant here, as it is about the other direction (converting nullable to `not_null`), but I'm pretty sure a similar (but a bit more contrived) example could be constructed for this direction too.

But of course you see that permitting the other direction (implicit constructors, and explicit constructors that are not explicit enough) tears a hole in the static non-nullability guarantee that we strive to establish. Implicit conversion operators do not. So whatever contrived scenario may exist, it will be categorically less dangerous than implicit constructors.


> Being able to simply pass a `not_null<T>` as a `T` to some function seems useful at first, but after introducing `not_null` in our codebase, I found that most things wrapped once in `not_null` basically never get unwrapped. Almost every time we had to use `as_nullable` (because of GCC 4.7.3) to pass a `not_null<T>` as a `T`, we discovered an interface that could not handle NULLs, so it should have consumed a `not_null<T>` in teh first place, eliminating the need for the conversion. This way, `as_nullable` became a marker for the borders between the "GSL-ed" and "non-GSL-ed" parts of the codebase; the remaining few cases where the consumer _could_ handle NULLs, but we wanted to pass a `not_null` were simply commented.

Thank you for sharing this experience. I appreciate field reports like this, not least because I have had only limited use for `not_null<>` myself. I'll reconsider this when it's time for version 1, and ideally after I've gotten more input on the impact of this change on other codebases.

> (But then I have to modify the test suite to include tests with both values of the config. :D)

The main challenge is to do this without doubling the build time of the CI pipeline :)

**Comment by petamas** [comment]
> But of course you see that permitting the other direction (implicit constructors, and explicit constructors that are not explicit enough) tears a hole in the static non-nullability guarantee that we strive to establish. Implicit conversion operators do not. So whatever contrived scenario may exist, it will be categorically less dangerous than implicit constructors.

OK, you caught me. :D I'll have to concede on this to you.

> The main challenge is to do this without doubling the build time of the CI pipeline :)

That shouldn't be that hard in theory; we only need to build `not_null.t.cpp` with the alternate configuration. Doing that in practice may need some CMake mojo, but I'll try my best. (Up until now, I could get away not really learning CMake, but I see that I'll have to sooner or later, for this and the "exceptionless" variants, and also for my job.)

> With "upstream" I meant to refer to the Core Guidelines maintainers. Like many parts of GSL, not_null<> is severely under-specified.

I think I'm severely underqualified for that kind of discussion, but if/when you start it, I'll read it for sure, and try to contribute if I can.

**Comment by mbeutel** [comment]
I'll close this issue; if you want to introduce a config switch for suppressing implicit conversion operators on `not_null<>`, I'm happy to accept a PR.
